ci/cloudbuild/cloudbuild.yaml

```diff
@@ -142,3 +142,38 @@ steps:
         xargs -r printf "gcr.io/${PROJECT_ID}/${_IMAGE}@%s\n" | \
         xargs -r -P 4 -L 32 gcloud container images delete -q --force-delete-tags
       exit 0
+
+  # Cancels *this* specific build one is already queued up for a newer commit.
+  # This frees up resources to run the build for the newer commit.
+- name: 'gcr.io/google.com/cloudsdktool/cloud-sdk'
+  waitFor: [ '-' ]
+  entrypoint: 'bash'
+  args:
+    - '-c'
+    - |
+      set -xe
+      ctime="$(gcloud builds describe --format 'value(create_time)' '${BUILD_ID}')"
+      query="tags=pr"
+      query+=" AND tags=${_PR_NUMBER}"
+      query+=" AND tags=${_BUILD_NAME}"
+      query+=" AND substitutions.COMMIT_SHA != ${COMMIT_SHA}"
+      query+=" AND create_time > ${ctime}"
+      gcloud builds list --limit 1 --format='value(id)' --filter "${query}" | \
+        grep . && gcloud builds cancel '${BUILD_ID}' || exit 0
+
+  # Cancels in-progress builds for previous commits in the current PR so we can
```

Comment:
This one took more thinking to convince myself it was safe.  You are cancelling *all* the builds for the previous SHA.  That might include a build that does not yet have a replacement, but presumably the replacement will be created shortly?

Reply:
yes, that's the idea. Actually, it cancels all builds for _all_ previous SHAs, which could be more than one. Yes, even if a replacement build doesn't exist yet, one should be created.

My reasoning was that once a new commit is pushed, we really don't care about any builds on the previous commits, so cancel 'em all.

Reply:
Also, I did consider only cancelling old instances of the current build. For example, the `gcc-58-pr` build would only cancel old instances of that particular build. In a way, that seems safer.

The reason I didn't do that was because we hit the GCB max quota limit on every PR push, so many newer builds wouldn't even be scheduled to run until the in-progress older instances finished and freed up a slot. So instead, this allows _any_ build from the "new commit" to cancel _all_ of the builds for older commits to hopefully free up more quota quickly.

Note: If there was no quota limitation, I probably wouldn't have done it this way.